Set `ChannelUpdate` `htlc_maximum_msat` using the peer's value

ViktorTigerstrom opened the pull request:
Closes #1443

Use the `counterparty_max_htlc_value_in_flight_msat` value, and not the
`holder_max_htlc_value_in_flight_msat` value when creating the
`htlc_maximum_msat` value for `ChannelUpdate` messages.

BOLT 7 specifies that the field MUST be less than or equal to
`max_htlc_value_in_flight_msat` received from the peer, which we
currently are not guaranteed to adhere to by using the holder (our) value.

https://github.com/lightning/bolts/blob/cf4fddd99eae9f15a527dfb785b47cd3c625c0fd/07-routing-gossip.md?plain=1#L469

TheBlueMatt reviewed (commented):
LG, needs a test, may be easiest to make the in-flight limit a config option while you're at it.

ViktorTigerstrom commented:
> LG, needs a test, may be easiest to make the in-flight limit % a config option while you're at it.

Sure sounds good, I'll fix that! Won't have time to address that until this weekend though unfortunately, I hope that's ok :)

vincenzopalazzo reviewed (dismissed):
LGTM

ViktorTigerstrom commented on lightning/src/ln/channel.rs:
```diff
@@ -5855,13 +5874,14 @@ impl<Signer: Sign> Writeable for Channel<Signer> {
 			{ Some(self.holder_selected_channel_reserve_satoshis) } else { None };
 		let serialized_holder_htlc_max_in_flight =
-			if self.holder_max_htlc_value_in_flight_msat != Self::get_holder_max_htlc_value_in_flight_msat(self.channel_value_satoshis)
+			if self.holder_max_htlc_value_in_flight_msat != Self::get_holder_max_htlc_value_in_flight_msat(self.channel_value_satoshis, self.config)
```
Note:
For this to work, this assumes that the channel config will remain not being possible to update for channels which are already created. If we in the future make the channel config updatable for live channels, this will start setting a `Some` value for `serialized_holder_htlc_max_in_flight` for channels where the `holder_max_htlc_value_in_flight_msat_channel_value_percent` has been changed, which will break backwards compatibility for those channels.

If this is a problem, please let me know :)

ViktorTigerstrom reviewed (commented):
Sorry for the delay in continuing this PR! Added a few comments which may be helpful when reviewing it.

If you think that it's more suitable that I create a new PR with the 2 first commits, I can of course do that!

TheBlueMatt commented on lightning/src/util/config.rs:
```diff
@@ -269,6 +269,13 @@ pub struct ChannelConfig {
+	/// Should be set to a value between 1-100, where the value signals the percantage to be used.
+	///
+	/// Default value: 10.
+	pub holder_max_htlc_value_in_flight_msat_channel_value_percent: u8,
```
Maybe `max_inbound_htlc_value_in_flight_percent_of_channel` - I don't like the `holder` terminology so much in public things like this if we can avoid it "inbound_htlc" seems more clear because, well, its an inbound HTLC. That said, I don't really care if the name is long, as long as its descriptive.

TheBlueMatt replied:
Also this should be in `ChannelHandshakeConfig`.

ViktorTigerstrom replied:
Ok! Given that we don't serialize the `ChannelHandshakeConfig` in the `Channel` struct, should I implement the channel serialization as follows:
- If the `max_inbound_htlc_value_in_flight_percent_of_channel` has been modified to any other value than the default, we write the `holder_max_htlc_value_in_flight_msat` into the tlv fields (as it's been configured), despite that it will break backwards compatibility?

TheBlueMatt commented on lightning/src/util/config.rs:
```diff
@@ -269,6 +269,13 @@ pub struct ChannelConfig {
+	/// Sets how many the percent of the channel value the holder_max_htlc_value_in_flight_msat
+	/// will be assigned as.
+	///
+	/// Should be set to a value between 1-100, where the value signals the percantage to be used.
```
What happens if the value is *not* within that range? Do we panic? Do we simply bound it to that range, etc.

TheBlueMatt commented on lightning/src/util/config.rs:
```diff
@@ -269,6 +269,13 @@ pub struct ChannelConfig {
 	/// [`Normal`]: crate::chain::chaininterface::ConfirmationTarget::Normal
 	/// [`Background`]: crate::chain::chaininterface::ConfirmationTarget::Background
 	pub force_close_avoidance_max_fee_satoshis: u64,
+	/// Sets how many the percent of the channel value the holder_max_htlc_value_in_flight_msat
```
Need *way* more docs here - without knowing anything about the lightning protocol, a user reading this documentation should be able to identify what the imlications of setting this value higher or lower will be.

It should probably link to `ChannelConfig::cltv_expiry_delta` and `ChannelHandshakeConfig::our_to_self_delay`, noting that this limits the HTLC exposure for inbound HTLCs only, It may also note that there is currently no ability to limit the outbound HTLC exposure.

ViktorTigerstrom replied:
Ok thanks, will update!

TheBlueMatt commented on lightning/src/ln/channel.rs:
```diff
@@ -865,6 +870,13 @@ impl<Signer: Sign> Channel<Signer> {
 		}
 
+		// Check that config has valid values set
+		if !Self::config_htlc_in_flight_value_is_valid(config.channel_options){
```
Lets just bound the value instead of failing here, no?

ViktorTigerstrom replied:
Do you mean that if the user sets a value that's `< 1`,  we'll set the `holder_max_htlc_value_in_flight_msat` using `1%` when creating the channel (and `100%` if `> 100`)?

Or do you mean that we should strictly bind the `ChannelHandshakeConfig` to only allow a `max_inbound_htlc_value_in_flight_percent_of_channel` value between 1-100? 

A strict binding would be a much nicer solution if there is an easy way to do that in Rust, but as I'm still quite new to Rust, I'm not aware an easy solution/type for that, and my searches yielded no great results. I was also a bit afraid if such a solution would create any implications for the language bindings, which is why I opted for dealing with it in the `Channel` struct instead. If you have any tips of how to solve this though, I'd be very thankful for those!

TheBlueMatt replied:
> Do you mean that if the user sets a value that's < 1, we should set the holder_max_htlc_value_in_flight_msat using 1% when creating the channel (and 100% if > 100)?

This. I think the "most" rust-y way to do this would be to check the value when constructing the `ChannelHandshakeConfig`, but I dunno that we really want to go that far here - there's value, at least in the bindings, to just having a POD struct here. The docs should be really clear that it needs to be between 1-100, but if users set it to something crazy we can just bound it for them and use 1 or 100.

TheBlueMatt reviewed (commented):
Nice! Thanks for tackling.

ViktorTigerstrom commented:
Addressed the latest feedback! I wasn't completely sure what you wanted me to mention about `ChannelConfig::cltv_expiry_delta` and `ChannelHandshakeConfig::our_to_self_delay `, so if I understood you incorrectly, please let me know :)!

Also, let me know if I should rebase on upstream as there are now a small conflict.

Additionally, let me know if I should modify this test function to use the new config option instead of directly modifying the `Channel::holder_max_htlc_value_in_flight_msat` value :) :
https://github.com/lightningdevkit/rust-lightning/blob/f53d13bcb8220b3ce39e51a4d20beb23b3930d1f/lightning/src/ln/functional_tests.rs#L138

ViktorTigerstrom commented:
Thanks a lot for the detailed feedback on the documentation @TheBlueMatt! Updated the docs according to the feedback. There were now 3 "Note that" statements in a row, so made it into a list instead to make the text flow better. In case you don't like that idea I'll of course change it back :).

Squashed the previous commits, and pushed the new changes in a seperate fixup commit!

valentinewallace commented on lightning/src/ln/channel.rs:
```diff
@@ -803,9 +807,23 @@ macro_rules! secp_check {
-	fn get_holder_max_htlc_value_in_flight_msat(channel_value_satoshis: u64) -> u64 {
-		channel_value_satoshis * 1000 / 10 //TODO
+	/// Returns the value to use for `holder_max_htlc_value_in_flight_msat` as a percentage of the
+	/// `channel_value_satoshis` in msat, based on how many of the percent its been configured to
+	/// be set to by the user.
```
s/based on how many of the percent its been configured to be set to by the user/[`ChannelHandkshakeConfig::max_inbound_htlc_value_in_flight_percent_of_channel`] (double check that link lol)

ViktorTigerstrom replied:
Thanks, edited it to:
set through [`ChannelHandkshakeConfig::max_inbound_htlc_value_in_flight_percent_of_channel`] 
But can remove the "set through" if you prefer :)!

valentinewallace commented on lightning/src/ln/channel.rs:
```diff
@@ -803,9 +807,23 @@ macro_rules! secp_check {
+	/// `channel_value_satoshis` in msat, based on how many of the percent its been configured to
+	/// be set to by the user.
+	///
+	/// As we only allow the user to set the percentage to a value between `1-100`, if the user
+	/// sets a value less than `1`, the function will result in `1` percent of the
+	/// `channel_value_satoshis`, as well as `100` percent if the user has set a value larger
+	/// than `100`.
```
Can replace this paragraph with: "The effective percentage is lower bounded by 0% and upper bounded by 100%."

ViktorTigerstrom replied:
Thanks! I figured you probably ment "lower bounded by 1%", so I edited to that, but let me know if I misunderstood :).

valentinewallace reviewed (commented):
Still going through the documentation, but the tests and code look good!

ViktorTigerstrom commented:
Thanks a lot for the review @valentinewallace! Addressed your suggestion in the latest commit!

ViktorTigerstrom commented on lightning/src/util/config.rs:
```diff
@@ -48,6 +48,30 @@ pub struct ChannelHandshakeConfig {
+	///
+	/// Note that:
+	/// * If configured to another value than the default value 10, any new channels
+	/// created with the non default value will cause versions of LDK prior to 0.0.100 to refuse to
```
Hmm could you please check the part about "LDK prior to 0.0.100" here? The `holder_max_htlc_value_in_flight_msat` was added to the TLV fields in LDK 0.0.104 ( f69311c ). 

As I'm still a bit uncertain about our serialization:
As the field was added with an is even number (`6`), does that mean that versions prior to LDK 0.0.103 will also fail when deserializing the `Channel` (hence this should be changed)?

ViktorTigerstrom replied:
Thanks for clarifying :)

ViktorTigerstrom commented:
Sorry, fixed a small spelling mistake in the latest commit.

valentinewallace commented on lightning/src/util/config.rs:
```diff
@@ -48,6 +48,30 @@ pub struct ChannelHandshakeConfig {
+	/// * This caps the total value for inbound HTLCs in-flight only, and there's currently
+	/// no way to configure the cap for the total value of outbound HTLCs in-flight.
+	///
+	/// * The online requirements for ensuring the safety of HTLC-encumbered funds are
+	/// different from the non-HTLC-encumbered funds. This makes this an important knob to restrict
+	/// exposure to loss due to being offline for too long.
+	/// See [`ChannelHandshakeConfig::our_to_self_delay`] and [`ChannelConfig::cltv_expiry_delta`]
+	/// for more information.
```
suggestion: s/The online requirements for ensuring/The requirements for your node being online to ensure (just wanna make sure it's clear)

valentinewallace reviewed (commented):
Think I'm ACK after these last minor comments

ViktorTigerstrom commented:
Addressed the latest feedback, and updated the commit message for the first commit.

I squashed the fixup commits now as the new changes were quite minor. I hope that's ok :)
